File: frontend/src/api/paySchedule.ts
import { apiFetch } from './client'

export type Frequency = 'weekly' | 'biweekly' | 'semimonthly' | 'monthly'
export type WeekendRule = 'none' | 'previous_business_day' | 'next_business_day'

export type Schedule = {
  id: number
  frequency: Frequency
  effective_from: string
  anchor_date: string | null
  day_of_month_1: number | null
  day_of_month_2: number | null
  weekend_rule: WeekendRule
  notes: string | null
}

export type ScheduleInput = {
  frequency: Frequency
  effective_from: string
  anchor_date?: string | null
  day_of_month_1?: number | null
  day_of_month_2?: number | null
  weekend_rule: WeekendRule
  notes?: string | null
}

export type PreviewPeriod = {
  start_date: string
  end_date: string
  is_transition: boolean
  days: number
}

export type Preview = {
  first_pay_date: string
  periods: PreviewPeriod[]
  transition: PreviewPeriod | null
  unshifted_dates: string[]
}

export type PayPeriod = {
  id: number
  start_date: string
  end_date: string
  schedule_id: number
  is_transition: boolean
  days: number
}

export function fetchSchedule(signal?: AbortSignal): Promise<{
  current: Schedule | null
  history: Schedule[]
}> {
  return apiFetch('/pay-schedule', { signal })
}

export function previewSchedule(body: ScheduleInput): Promise<Preview> {
  return apiFetch<Preview>('/pay-schedule/preview', { method: 'POST', body })
}

export function commitSchedule(body: ScheduleInput): Promise<Schedule> {
  return apiFetch<Schedule>('/pay-schedule', { method: 'POST', body })
}

export function fetchPeriods(
  range: { from?: string; to?: string } = {},
  signal?: AbortSignal,
): Promise<{ items: PayPeriod[] }> {
  const params = new URLSearchParams()
  if (range.from) params.set('from', range.from)
  if (range.to) params.set('to', range.to)
  const query = params.toString()
  return apiFetch(`/pay-periods${query ? `?${query}` : ''}`, { signal })
}

export function fetchCurrentPeriod(signal?: AbortSignal): Promise<PayPeriod> {
  return apiFetch<PayPeriod>('/pay-periods/current', { signal })
}

File: backend/app/schemas/pay_schedule.py
"""Pay schedule and pay period schemas."""

from datetime import date
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

Frequency = Literal["weekly", "biweekly", "semimonthly", "monthly"]
WeekendRule = Literal["none", "previous_business_day", "next_business_day"]


class ScheduleInput(BaseModel):
    frequency: Frequency
    effective_from: date
    anchor_date: date | None = None
    day_of_month_1: int | None = Field(default=None, ge=1, le=31)
    day_of_month_2: int | None = Field(default=None, ge=1, le=31)
    weekend_rule: WeekendRule = "none"
    notes: str | None = Field(default=None, max_length=500)


class ScheduleOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    frequency: Frequency
    effective_from: date
    anchor_date: date | None
    day_of_month_1: int | None
    day_of_month_2: int | None
    weekend_rule: WeekendRule
    notes: str | None


class ScheduleHistoryOut(BaseModel):
    current: ScheduleOut | None
    history: list[ScheduleOut]


class PeriodOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    start_date: date
    end_date: date
    schedule_id: int
    is_transition: bool
    days: int = 0


class PeriodListOut(BaseModel):
    items: list[PeriodOut]


class PreviewPeriodOut(BaseModel):
    start_date: date
    end_date: date
    is_transition: bool
    days: int


class PreviewOut(BaseModel):
    first_pay_date: date
    periods: list[PreviewPeriodOut]
    transition: PreviewPeriodOut | None
    #: Pay dates where the weekend rule was skipped to keep the timeline in order.
    unshifted_dates: list[date]
